src/workerd/api/web-socket.c++

```diff
@@ -12,6 +12,22 @@
 
 namespace workerd::api {
 
+kj::Maybe<WebSocketProtocolError> WebSocketProtocolError::fromException(const kj::Exception& ex) {
+  auto maybeContext = ex.getContext();
+  while (maybeContext != nullptr) {
+    auto& context = KJ_ASSERT_NONNULL(maybeContext);
+    if (magicFileValue == context.file) {
+      return WebSocketProtocolError(context.line, kj::str(context.description));
+    }
+    maybeContext = context.next;
+  }
+  return nullptr;
+}
+
+void WebSocketProtocolError::encodeToException(kj::Exception &ex) && {
```

Comment:
This might be a little overengineered, but since encodeToException moves `description` into the exception, you can't usefully call encodeToException twice on the same `WebSocketProtocolError` object; the second call would put an empty description in there.

The ref qualifier means you have to type `kj::mv(wspe).encodeToException`, which is a big hint to the programmer that they shouldn't use `wspe` any more after that.

The thought process was something like this:

1. I'll just copy `description` into the exception.
2. Oh no, it's a `kj::String` so copying is ugly and reviewers won't like it. I can still move `description`, though.
3. Well, now after I call `wspe.encodeToException`, the object isn't usable any more, but there's no way to tell. That could confuse other people or future me.
4. I can toss a ref qualifier on `encodeToException`; then the caller will have to type `kj::mv(wspe)`, providing a clue that `wspe` should no longer be used for anything.

Reply:
I think the moving of `description` is fine; `kj::String` is movable, and it just winds up as `""` after being moved from.

I'm more thinking about code like this:

```
WebSocketProtocolError wspe{1009, "message too big"};
kj::Exception ex1, ex2;

wspe.encodeToException(ex1);  // works as expected, but "uses up" wspe
wspe.encodeToException(ex2);  // oops, ex2 now has partial data in it, but this is not obvious
```

The ref qualifier is there so you instead have to type

```
kj::mv(wspe).encodeToException(ex1);  // This is fine
kj::mv(wspe).encodeToException(ex2);  // Wait, what? We shouldn't use wspe after moving from it on the previous line. Something's fishy here.
```

Reply:
Ah I see, this stops you from doing something wrong at (or even before) compile time, whereas if you were to do `kj::Maybe<kj::String> description` and fail the op if it's null, you find out at runtime (at which point you hope you have test coverage). Makes sense, thanks for the details!